### PR TITLE
Text

### PR DESCRIPTION
**Pull request by za-ek**

### CODE REVIEW

**Comment by meyfa** [comment]
Thanks so much for contributing! Text has long been a missing feature. I'll review your code as quickly as possible — most likely this weekend.

File: src/Rasterization/Renderers/SVGRenderer.php [hunk 1]
```diff
@@ -57,6 +50,13 @@ public function render(SVGRasterizer $rasterizer, array $options, SVGNode $conte
 
             $this->renderStroke($image, $params, $stroke, $strokeWidth);
         }
+
+        $fill = $context->getComputedStyle('fill');
+        if (isset($fill) && $fill !== 'none') {
+            $fill = self::prepareColor($fill, $context);
+
+            $this->renderFill($image, $params, $fill);
+        }
```
Comment (meyfa) [inline src/Rasterization/Renderers/SVGRenderer.php]: To cite the SVG 2 specification, § 3.7.1, third paragraph:

> The order in which fill, stroke and markers are painted is determined by the paint-order property. The default is that fill is painted first, then the stroke, and then the marker symbols. [...]

As such, changing the paint order here is incorrect.

This also means the "renderer-rect-stroke-fill.png" test image should not be replaced.

Comment (meyfa) [reply src/Rasterization/Renderers/SVGRenderer.php]: Of course, I understand that doing it this way is required for rendering text with `imagettftext`. We need to come up with a solution to that.

File: src/Rasterization/Renderers/SVGTextRenderer.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+namespace SVG\Rasterization\Renderers;
+
+use SVG\Rasterization\SVGRasterizer;
+
+class SVGTextRenderer extends SVGRenderer
+{
+    protected function prepareRenderParams(SVGRasterizer $rasterizer, array $options)
+    {
+        $options['x'] += $rasterizer->getOffsetX();
+        $options['y'] += $rasterizer->getOffsetY();
```
Comment (meyfa) [inline src/Rasterization/Renderers/SVGTextRenderer.php]: Here, viewBox scaling and also output image scaling are not taken into account.

- x/y should be scaled in addition to the offset
- size should be scaled

File: src/Nodes/Texts/SVGText.php [hunk 3]
```diff
@@ -1,13 +1,83 @@
...
+    public function __construct($text = '', $x = 0, $y = 0)
+    {
+        parent::__construct();
+        $this->setValue($text);
+        if($x > 0) {
+            $this->setAttribute('x', $x);
+        }
+        if($y > 0) {
+            $this->setAttribute('y', $y);
+        }
```
Comment (meyfa) [inline src/Nodes/Texts/SVGText.php]: Why can text not start outside the viewport? I think that would be a perfectly valid thing, and these conditionals should not be here.

File: src/Nodes/Texts/SVGText.php [hunk 4]
```diff
@@ -1,13 +1,83 @@
...
+    }
+
+    public function rasterize(SVGRasterizer $rasterizer)
+    {
+        $rasterizer->render('text', array(
+            'x'         => $this->getAttribute('x'),
+            'y'         => $this->getAttribute('y'),
+            'size'      => intval($this->getComputedStyle('font-size')),
```
Comment (meyfa) [inline src/Nodes/Texts/SVGText.php]: Just calling `intval` might be a bit too easy. Font size can be specified in all kinds of units.

File: src/Rasterization/Renderers/SVGTextRenderer.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    protected function renderStroke($image, array $params, $color, $strokeWidth)
+    {
+        $x = $params['x'];
+        $y = $params['y'];
+        $px = $strokeWidth;
+
+        for ($c1 = ($x-abs($px)); $c1 <= ($x+abs($px)); $c1++) {
+            for ($c2 = ($y - abs($px)); $c2 <= ($y + abs($px)); $c2++) {
+                imagettftext($image, $params['size'], 0, $c1, $c2, $color, $params['font_path'], $params['text']);
```
Comment (meyfa) [inline src/Rasterization/Renderers/SVGTextRenderer.php]: The nested loops start to break down for stroke colors with alpha components. Then again, we have _no_ support right now, so this is still a lot better.

File: src/Nodes/Texts/SVGText.php (hunk 4 above)
Comment (stelgenhof) [reply src/Nodes/Texts/SVGText.php]: In addition, intval is a PHP4 function. Typecasting would be more performant in this case.

**Comment by meyfa** [comment]
Any updates on this? This is valuable work you did there, @za-ek!

But of course, if you don't want to or cannot spend time right now to continue this, that's fine. In that case, I will merge the code into a feature branch and apply the changes mentioned above, finally merging it into master.

Please respond with whether you'd like to continue yourself or not.

**Comment by za-ek** [comment]
There are 2 issues in codeclimate. I never used it so if you can please help me to solve this
https://codeclimate.com/github/meyfa/php-svg/pull/54

**Comment by meyfa** [comment]
@za-ek Don't worry about CodeClimate.

The code looks good now! In fact, way better than I could wish for. I see you even implemented `paint-order`. Thank you so much, great work!

Merging this now.

**Comment by meyfa** [comment]
This PR is released as part of [v0.9.0](https://github.com/meyfa/php-svg/releases/tag/v0.9.0).

**Pull request merged**